### DISCUSSION TITLE
CoW permute_axes

### DISCUSSION THREAD

**Discussion by emmatyping** [rust-ndarray/ndarray#1026] [Q&A] [answered]
I believe like `to_shape` (which is great btw, I really like using it!) it would be really handy to have a copy on write version of permute_axes, which only copies the matrix if it is written to. Would such a method be possible? I'm using `permute_axes` to do some tensor computations at the moment.

**Comment by jturner314** [comment, marked as answer]
I think what you're asking for is already available. [`CowArray`](https://docs.rs/ndarray/0.15.2/ndarray/type.CowArray.html) and [`ArcArray`](https://docs.rs/ndarray/0.15.2/ndarray/type.ArcArray.html) have copy-on-write behavior, and `.permuted_axes()` doesn't require a copy of the data, so you can just create a `CowArray` or `ArcArray` and permute its axes. Here's an example using `CowArray`:

```rust
use ndarray::prelude::*;

fn main() {
    let a = array![[0., 1., 2.], [3., 4., 5.]];
    assert_eq!(a.shape(), &[2, 3]);
    
    let mut b = CowArray::from(a.view()).permuted_axes([1, 0]);
    assert_eq!(b.shape(), &[3, 2]);
    assert_eq!(b, array![[0., 3.], [1., 4.], [2., 5.]]);
    
    assert!(b.is_view());
    b[[0, 0]] = 6.;
    assert!(b.is_owned());
    
    assert_eq!(a, array![[0., 1., 2.], [3., 4., 5.]]);
    assert_eq!(b, array![[6., 3.], [1., 4.], [2., 5.]]);
}
```

Does that meet your needs?
